File: java/src/main/structural/facade/Facade.java
package structural.facade;

// Facade: Provides a simplified interface to the complex subsystem
public class Facade {
    private final SubsystemA subsystemA;
    private final SubsystemB subsystemB;
    private final SubsystemC subsystemC;

    public Facade() {
        this.subsystemA = new SubsystemA();
        this.subsystemB = new SubsystemB();
        this.subsystemC = new SubsystemC();
    }

    public void simpleOperation() {
        System.out.println("Facade: Coordinating operations...");
        subsystemA.operationA();
        subsystemB.operationB();
        subsystemC.operationC();
        System.out.println("Facade: Operation completed.");
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Facade facade = new Facade();
//         
//         // The client interacts with the Facade instead of the complex subsystem
//         facade.simpleOperation();
//
//         // Expected Output:
//         // Facade: Coordinating operations...
//         // SubsystemA: Performing operation A.
//         // SubsystemB: Performing operation B.
//         // SubsystemC: Performing operation C.
//         // Facade: Operation completed.
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Overcomplicating the Facade:
//    - Mistake: Adding unnecessary logic inside the Facade, making it a bottleneck.
//    - Fix: Keep the Facade focused on delegating tasks to subsystems.
//
// 2. Ignoring Direct Access to Subsystems:
//    - Mistake: Clients still use subsystems directly, defeating the purpose of the Facade.
//    - Fix: Encourage all interactions through the Facade.
//
// 3. Breaking Encapsulation:
//    - Mistake: Exposing internal details of the subsystems in the Facade.
//    - Fix: Keep the Facade's interface simple and hide implementation details.
//
// 4. Making Facade a Monolithic Class:
//    - Mistake: Facade grows too large and becomes difficult to maintain.
//    - Fix: Split responsibilities into multiple smaller facades if needed.

File: python/src/structural/facade/facade.py
from src.structural.facade.subsystem_a import SubsystemA
from src.structural.facade.subsystem_b import SubsystemB


class Facade:
    def __init__(self, subsystem_a: SubsystemA = None,
                 subsystem_b: SubsystemB = None):
        self._subsystem_a = subsystem_a or SubsystemA()
        self._subsystem_b = subsystem_b or SubsystemB()

    def operation(self) -> str:
        results = []
        results.append("Facade initializes subsystems:")
        results.append(self._subsystem_a.operation_a())
        results.append("Facade orders subsystems to perform the action:")
        results.append(self._subsystem_b.operation_b())
        return "".join(results)

# Example Usage:
# if __name__ == "__main__":
#     facade = Facade()
#     print(facade.operation())
#
# Expected Output:
# Facade initializes subsystems:
# SubsystemA: Ready
# Facade orders subsystems to perform the action:
# SubsystemB: Go!

# Common Mistakes to Avoid:
#
# 1. Violating Encapsulation:
#    - Mistake: Exposing too much of the subsystem through the facade.
#    - Fix: Only expose the necessary operations.
#
# 2. Overcomplicating the Facade:
#    - Mistake: Adding logic that belongs to subsystems.
#    - Fix: Keep the facade lightweight and delegate work.
#
# 3. Using Facade as God Object:
#    - Mistake: Making the facade do everything.
#    - Fix: Keep responsibilities separate and only orchestrate interactions.
#
# 4. Tight Coupling:
#    - Mistake: Hardcoding specific subsystem behavior into the facade.
#    - Fix: Inject subsystems if custom behavior is needed (as in the constructor).
